File: src/cmcandy/Java_Language_Answers/_0041.java
package cmcandy.Java_Language_Answers;

/**
 * Created by CHEN on 2020/5/14.
 */
public class _0041 {

    static class Solution {
        public int firstMissingPositive(int[] nums) {


            int len = nums.length - 1;
            int res = 0;
            int tmp = 0;
            while (res <= len) {

                //表示在正确的位置上，不用替换
                if (nums[res] == res + 1) {
                    res++;
                } else {
                    tmp = nums[res];
                    //nums[tmp-1] == tmp? tmp位置的值已经为tmp 可以不用替换，不加的话，碰到两个一样的值会陷入死循环的情况
                    if (tmp > len + 1 || tmp < res + 1 || nums[tmp - 1] == tmp) {
                        nums[res] = nums[len--];
                    } else {
                        //需要吧tmp-1位置上的数交换到
                        nums[res] = nums[tmp - 1];
                        nums[tmp - 1] = tmp;
                    }

                }

            }


            return res + 1;
        }
    }
        public static void main(String[] args) {

        }


    }

File: src/cmcandy/C_Language_Answers/_0041.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace ::std;
class Solution
{
public:
    int firstMissingPositive(vector<int> &nums)
    {

        int len = nums.size();
        if (len == 0 || (len == 1 && nums[0] != 1))
            return 1;
        else if (len == 1 && nums[0] == 1)
            return 2;
        int i = 0;
        while (i < len)
        {
            cout << nums[0] << nums[1] << nums[2]<<nums[3] << endl;
            if (nums[i] != i)
            {
                if (nums[i] >= len || nums[i] <= 0)
                {
                    nums[i] = 0;
                    i++;
                }
                else if (nums[nums[i]] != nums[i])
                {
                    int tmp = nums[nums[i]];
                    nums[nums[i]] = nums[i];
                    nums[i] = tmp;
                }else i++;
            }
            else
                i++;
        }
        for (int j = 1; j < len; j++)
        {
            if (nums[j] == 0)
                return j;
        }
        return i;
    }
};
int main()
{
    Solution s;
    int a[] = { 1, 1};
    vector<int> v(a, a + 4);
    s.firstMissingPositive(v);
}